operator/src/main/java/org/keycloak/operator/v2alpha1/WatchedSecretsStore.java

```diff
@@ -0,0 +1,237 @@
+        });
+    }
+
+    @Override
+    protected Optional<HasMetadata> getReconciledResource() {
+        Secret secret = existingStore != null ? existingStore : getNewStore();
+        secret.setData(null);
+        secret.setStringData(currentVersions);
+
+        return Optional.of(secret);
+    }
+
+    @Override
+    protected void setDefaultLabels(HasMetadata resource) {
+        super.setDefaultLabels(resource);
+        resource.getMetadata().getLabels().put(Constants.TYPE_LABEL, TYPE);
+    }
+
+    @Override
+    public void createOrUpdateReconciled() {
+        super.createOrUpdateReconciled();
+        addLabelsToWatchedSecrets();
+    }
+
+    public void addLabelsToWatchedSecrets() {
+        for (Secret secret : currentSecrets) {
+            if (secret.getMetadata() == null
+                    || secret.getMetadata().getLabels() == null
+                    || !secret.getMetadata().getLabels().containsKey(Constants.TYPE_LABEL)) {
+
+                Log.infof("Adding label to Secret \"%s\"", secret.getMetadata().getName());
+
+                secret = new SecretBuilder(secret)
+                        .editMetadata()
+                        .addToLabels(Constants.TYPE_LABEL, TYPE_WATCHED)
+                        .endMetadata()
+                        .build();
+
+                client.secrets().patch(secret);
+            }
+        }
+    }
+
+    private Secret fetchExistingStore() {
+        return client.secrets().inNamespace(getNamespace()).withName(getName()).get();
+    }
+
+    private Secret getNewStore() {
+        return new SecretBuilder()
+                .withNewMetadata()
+                    .withName(getName())
+                    .withNamespace(getNamespace())
+                .endMetadata()
+                .build();
+    }
+
+    private Map<String, String> getNewLastObservedVersions() {
+        if (existingStore != null && existingStore.getData() != null) {
+            return existingStore.getData().entrySet().stream()
+                    .collect(Collectors.toMap(
+                            Map.Entry::getKey,
+                            e -> new String(Base64.getDecoder().decode(e.getValue()))
+                    ));
+        }
+        else {
+            return Collections.emptyMap();
+        }
+    }
+
+    private Map<String, String> getNewCurrentVersions() {
+        return currentSecrets.stream()
+                .collect(Collectors.toMap(s -> s.getMetadata().getName(), this::getSecretVersion));
+    }
+
+    private String getSecretVersion(Secret secret) {
+        String serializedData = Serialization.asYaml(secret.getData());
+        try {
+            // using hashes as it's more robust than resource versions that can change e.g. just when adding a label
+            byte[] bytes = MessageDigest.getInstance("MD5").digest(serializedData.getBytes(StandardCharsets.UTF_8));
+            return new BigInteger(1, bytes).toString(16);
```

Comment:
Do we need to convert it to a number?
can it be simply `new String(bytes)` ?

Reply:
We need to convert bytes to a hexadecimal string.